zanata-war/src/main/java/org/zanata/rest/editor/service/SuggestionsService.java

```diff
@@ -0,0 +1,142 @@
+
+        Either<LocaleId, Response> sourceLocaleOrError = getSourceLocale(sourceLocaleString);
+        if (sourceLocaleOrError.isRight()) {
+            return sourceLocaleOrError.right();
+        }
+        LocaleId sourceLocale = sourceLocaleOrError.left();
+
+        Either<LocaleId, Response> transLocaleOrError = getTransLocale(transLocaleString);
+        if (transLocaleOrError.isRight()) {
+            return transLocaleOrError.right();
+        }
+        LocaleId transLocale = transLocaleOrError.left();
+
+        List<Suggestion> suggestions = transMemoryService.searchTransMemoryWithDetails(transLocale, sourceLocale,
+                new TransMemoryQuery(query, searchType));
+
+        // Wrap in generic entity to prevent type erasure, so that an
+        // appropriate MessageBodyReader can be used.
+        // see docs for GenericEntity
+        GenericEntity<List<Suggestion>> entity = new GenericEntity<List<Suggestion>>(suggestions) {};
+
+        return Response.ok(entity).build();
+    }
+
+    private Either<LocaleId, Response> getSourceLocale(String localeString) {
+        return getLocaleOrError(localeString, "Unrecognized source locale: \"%s\"");
+    }
+
+    private Either<LocaleId, Response> getTransLocale(String localeString) {
+        return getLocaleOrError(localeString, "Unrecognized translation locale: \"%s\"");
+    }
+
+    /**
+     * Try to get a valid locale for a given string. If the string does not match a supported
+     * locale, generate an error response with the given error message.
+     *
+     * @param localeString used to look up the locale
+     * @param errorFormat error to display. Include "%s" to substitute the locale string.
+     * @return a LocaleId if the given string matches one, otherwise an error response.
+     */
+    private Either<LocaleId, Response> getLocaleOrError(final String localeString, final String errorFormat) {
+        final String errorMessage = String.format(errorFormat, localeString);
+        final Response.ResponseBuilder errorResponse = Response.status(BAD_REQUEST).entity(errorMessage);
+
+        try {
+            @Nullable HLocale hLocale = localeService.getByLocaleId(localeString);
+            if (hLocale == null) {
+                return Either.right(errorResponse.build());
+            }
+            return Either.left(hLocale.getLocaleId());
+        } catch (IllegalArgumentException e) {
+            return Either.right(errorResponse.build());
+        }
+    }
+
+
+    /**
+     * Try to get a valid search type constant for a given string. If the string does not
+     * match a valid search type, generate an error response with appropriate error message.
+     *
+     * @param searchTypeString used to look up the search type.
+     * @return A SearchType if the given string matches one, otherwise an error response.
+     */
+    private Either<SearchType, Response> getSearchType(String searchTypeString) {
```

Comment:
When searchTypeString is null will it trigger NPE or IllegalArgumentException? May need to catch both?
Alternatively, we can loop through SearchType.values() array and see if we can find a match *case insensitively*. This may offer friendlier user experience but just a suggestion.